lib/semantics/resolve-names.cc

```diff
@@ -3907,32 +3902,27 @@ bool DeclarationVisitor::CanBeTypeBoundProc(const Symbol &symbol) {
   }
 }
 
-void DeclarationVisitor::CheckExplicitInterface(Symbol &symbol) {
-  if (const auto *details{symbol.detailsIf<ProcEntityDetails>()}) {
-    if (const Symbol * interface{details->interface().symbol()}) {
-      if (!interface->HasExplicitInterface() && !context().HasError(symbol)) {
-        if (!context().HasError(*interface)) {
-          Say(symbol.name(),
-              "The interface of '%s' is not an abstract interface or a "
-              "procedure with an explicit interface"_err_en_US);
-        }
-        context().SetError(symbol);
-      }
-    }
+Symbol *DeclarationVisitor::NoteInterfaceName(const parser::Name &name) {
+  // The symbol is checked later by CheckExplicitInterface() to ensure
+  // that it defines an explicit interface.  The name can be a forward
+  // reference.
+  if (!NameIsKnownOrIntrinsic(name)) {
+    Resolve(name, MakeSymbol(InclusiveScope(), name.source, Attrs{}));
   }
+  return name.symbol;
 }
 
-Symbol *DeclarationVisitor::FindExplicitInterface(const parser::Name &name) {
-  auto *symbol{FindSymbol(name)};
-  if (!symbol) {
-    Say(name, "Explicit interface '%s' not found"_err_en_US);
-  } else if (!symbol->HasExplicitInterface()) {
-    SayWithDecl(name, *symbol,
-        "'%s' is not an abstract interface or a procedure with an"
-        " explicit interface"_err_en_US);
-    symbol = nullptr;
+void DeclarationVisitor::CheckExplicitInterface(Symbol &symbol) {
+  if (const Symbol * interface{FindInterface(symbol)}) {
+    const Symbol *subp{FindSubprogram(*interface)};
+    if (subp == nullptr || !subp->HasExplicitInterface()) {
+      Say(symbol.name(),
+          "The interface of '%s' (%s) is not an abstract interface or a "
+          "procedure with an explicit interface"_err_en_US,
+          symbol.name(), interface->name());
```

Comment:
I think the interface name should be in quotes too, since it's a name from the user program.